File: 0139-word-break/0139-word-break.cpp
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        
        // Initialize dp array
        const int N = s.size();
        vector<bool> dp(N+1, false);
        dp[N] = true;

        // Fill dp array
        for (int i = N - 1; i >= 0; i--) {

            // go through every word
            for (string& word : wordDict) {
                if (i + word.size() > N) { continue; }
                else if ((s.substr(i, word.size()) == word) && dp[i + word.size()]) {
                    dp[i] = true;
                    break;
                }
            }
        }
        return dp[0];
    }
};

File: 0139-word-break/0139-word-break.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        
        # Initialize dp array
        N = len(s)
        dp = [False] * (N+1)
        dp[N] = True
        
        
        # fill dp array
        for i in range(N-1, -1, -1):
            
            for word in wordDict:
                if i + len(word) > N:
                    continue
                elif s[i:i + len(word)] == word and dp[i + len(word)]:
                    dp[i] = True
                    break
        return dp[0]
                
                    
            
            
        